batch/batch/cloud/azure/driver/billing_manager.py

```diff
@@ -0,0 +1,103 @@
+from gear import Database, transaction
+
+from ....driver.billing_manager import CloudBillingManager, ResourceVersions, resource_prefix_version_to_name
+from ..resources import AzureVMResource, AzureDiskResource
+
+from .pricing import AzureVMPrice, AzureDiskPrice, fetch_prices
+
+log = logging.getLogger('resource_manager')
+
+
+AzureVMIdentifier = namedtuple('AzureVMIdentifier', ['machine_type', 'preemptible', 'region'])
+
+
+class AzureBillingManager(CloudBillingManager):
+    @staticmethod
+    async def create(db: Database,
+                     pricing_client: aioazure.AzurePricingClient,  # BORROWED
+                     regions: List[str],
+                     ):
+        rm = AzureBillingManager(db, pricing_client, regions)
+        await rm.refresh_resources()
+        await rm.refresh_resources_from_retail_prices()
+        return rm
+
+    def __init__(self,
+                 db: Database,
+                 pricing_client: aioazure.AzurePricingClient,
+                 regions: List[str],
+                 ):
+        self.db = db
+        self.resource_versions = ResourceVersions()
+        self.resource_rates: Dict[str, float] = {}
+        self.pricing_client = pricing_client
+        self.regions = regions
+        self.vm_price_cache: Dict[AzureVMIdentifier, AzureVMPrice] = {}
+
+    async def get_spot_billing_price(self, machine_type: str, location: str) -> float:
+        vm_identifier = AzureVMIdentifier(machine_type=machine_type, preemptible=True, region=location)
+        return self.vm_price_cache[vm_identifier].cost_per_hour
+
+    async def refresh_resources_from_retail_prices(self):
+        log.info('refreshing resources from retail prices')
+        resource_updates = []
+        resource_version_updates = []
+
+        for price in await fetch_prices(self.pricing_client, self.regions):
+            if isinstance(price, AzureVMPrice):
+                resource_prefix = AzureVMResource.generate_prefix(price.machine_type, price.preemptible, price.region)
+                latest_resource_rate = rate_instance_hour_to_fraction_msec(price.cost_per_hour, 1024)
+            else:
+                assert isinstance(price, AzureDiskPrice)
+                resource_prefix = AzureDiskResource.generate_prefix(price.disk_name, price.redundancy_type, price.region)
+                latest_resource_rate = rate_gib_month_to_mib_msec(price.cost_per_gib_month)
+
+            latest_resource_version = price.version
+            resource_name = resource_prefix_version_to_name(resource_prefix, latest_resource_version)
+
+            current_resource_rate = self.resource_rates.get(resource_name)
+            current_resource_version = self.resource_versions.latest_version(resource_prefix)
+
+            if current_resource_rate is None:
+                resource_updates.append((resource_name, latest_resource_rate))
+            elif abs(current_resource_rate - latest_resource_rate) >= 1E-20:
```

Comment:
In 2022, we need to sit down and think about why our rates are floating point numbers. This makes me deeply uncomfortable.

Reply:
Why can't this be equality? The latest resource rate and the current resource rate should absolutely be the exact same floating point number. How would they not be the same? Don't the latest and current both ultimately come from the database?